### ISSUE TITLE
What does "out_axes=" do in vmap?

### ISSUE THREAD

**Issue by LenCewa** [google/jax#1368]
Hello,
I don't get the use for the out_axes in vmap. For example if I modify your example to:

```
vv = lambda x, y: jnp.vdot(x, y)  # ([a], [a]) -> []
mv = vmap(vv, (0, None))  # ([a,b], [b]) -> [a]
mm = vmap(mv, (None, 1))  # ([a,b], [b,c]) -> [a,c]
```

I didn't specify any out_axes but I still get the same result. Switching the out_axes to:

```
vv = lambda x, y: jnp.vdot(x, y)  # ([a], [a]) -> []
mv = vmap(vv, (0, None), 1)  # ([a,b], [b]) -> [a]
mm = vmap(mv, (None, 1), 0)  # ([a,b], [b,c]) -> [a,c]
```

leaves the result unchanged too.

**Comment by xysun** [comment]
It does change things when the output shape has more than 1 axis. 

For example: 

```
vv = lambda x, y: jnp.vdot(x,y) # ([a],[a]) -> []
mv = vmap(vv, (0,None),1) # ([a,b], [b]) -> [a]; out_axis does not matter since output shape only has one axis
mm1 = vmap(mv, (None,1),0) # ([a,b], [b,c]) -> [c,a]; output transposed because input_axis is 1 and output_axis is 0
mm2 = vmap(mv, (None,1),1) # ([a,b],[b,c]) -> [a,c]; expected shape because input and output axis aligned. 
```

Concrete example:
```
a1 = np.array([[1,2,3],[1,2,3]]) # shape (2,3)
a2 = np.array([[1],[2],[3]]) # shape (3,1)
r1 = mm1(a1,a2) # shape (1,2)
r2 = mm2(a1,a2) # shape (2,1)
```

**Comment by LenCewa** [comment]
Thanks!

**Comment by hvater** [comment]
A very late addition to this topic:

`out_axes` in` vmap` is also helpful when you work with pytrees where one element is a scalar (or when it is not supposed to be mapped):

```
def dict_func(bar):
    a = bar["c"] * (bar["a"] * 4)
    b = bar["c"] * (bar["b"] - jnp.mean(bar["b"]))
    return dict(a=a, b=b, c=bar["c"])

bar = dict(
    a=jnp.array([[1, 3, 3], [0, 0, 1]]),
    b=jnp.array([[1, 2, 3], [0, 0, 1]]),
    c=0.5
)
```
Calling the function without `out_axes`, i.e.

```
jax.vmap(
    dict_func,
    in_axes=(dict(a=0, b=0, c=None),)
)(bar)
```
returns

```
{'a': Array([[2., 6., 6.],
        [0., 0., 2.]], dtype=float32, weak_type=True),
 'b': Array([[-0.5       ,  0.        ,  0.5       ],
        [-0.16666667, -0.16666667,  0.3333333 ]], dtype=float32),
 'c': Array([0.5, 0.5], dtype=float32, weak_type=True)}
```

Note that the 'c' entry is now an array which might not be your intention.

Calling

```
jax.vmap(
    dict_func,
    in_axes=(dict(a=0, b=0, c=None),),
    out_axes=(dict(a=0, b=0, c=None))
)(bar)
```
returns

```
{'a': Array([[2., 6., 6.],
        [0., 0., 2.]], dtype=float32, weak_type=True),
 'b': Array([[-0.5       ,  0.        ,  0.5       ],
        [-0.16666667, -0.16666667,  0.3333333 ]], dtype=float32),
 'c': 0.5}
```

which keeps the 'c' entry as a scalar.
